libbpf-tools/tcpconnlat.c

```diff
@@ -0,0 +1,209 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+// Copyright (c) 2020 Wenbo Zhang
+//
+// Based on tcpconnlat(8) from BCC by Brendan Gregg.
+// 11-Jul-2020   Wenbo Zhang   Created this.
+#include <argp.h>
+#include <arpa/inet.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <time.h>
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+#include "tcpconnlat.h"
+#include "tcpconnlat.skel.h"
+#include "trace_helpers.h"
+
+#define PERF_BUFFER_PAGES	16
+#define PERF_POLL_TIMEOUT_MS	100
+
+static struct env {
+	__u64 min_us;
+	pid_t pid;
+	bool timestamp;
+	bool verbose;
+} env;
+
+const char *argp_program_version = "tcpconnlat 0.1";
+const char *argp_program_bug_address = "<ethercflow@gmail.com>";
+const char argp_program_doc[] =
+"\nTrace TCP connects and show connection latency.\n"
+"\n"
+"USAGE: tcpconnlat [-h] [-t] [-p PID]\n"
+"\n"
+"EXAMPLES:\n"
+"    tcpconnlat              # summarize on-CPU time as a histogram"
+"    tcpconnlat 1            # trace connection latency slower than 1 ms"
+"    tcpconnlat 0.1          # trace connection latency slower than 100 us"
+"    tcpconnlat -t           # 1s summaries, milliseconds, and timestamps"
+"    tcpconnlat -p 185       # trace PID 185 only";
+
+static const struct argp_option opts[] = {
+	{ NULL, 'h', NULL, OPTION_HIDDEN, "Show the full help" },
+	{ "timestamp", 't', NULL, 0, "Include timestamp on output" },
+	{ "pid", 'p', "PID", 0, "Trace this PID only" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	static int pos_args;
+
+	switch (key) {
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'h':
+		argp_usage(state);
+		break;
+	case 'p':
+		errno = 0;
+		env.pid = strtol(arg, NULL, 10);
+		if (errno) {
+			fprintf(stderr, "invalid PID: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case 't':
+		env.timestamp = true;
+		break;
+	case ARGP_KEY_ARG:
+		if (pos_args++) {
+			fprintf(stderr,
+				"Unrecognized positional argument: %s\n", arg);
+			argp_usage(state);
+		}
+		errno = 0;
+		env.min_us = strtod(arg, NULL) * 1000;
+		if (errno || env.min_us <= 0) {
+			fprintf(stderr, "Invalid delay (in us): %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+		    const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
+{
+	const struct event *e = data;
+	char src[INET6_ADDRSTRLEN];
+	char dst[INET6_ADDRSTRLEN];
+	union {
+		struct in_addr  x4;
+		struct in6_addr x6;
+	} s, d;
+	static __u64 start_ts;
+
+	if (env.timestamp) {
+		if (start_ts == 0)
+			start_ts = e->ts_us;
+		printf("%-9.3f ", ((double)e->ts_us - start_ts) / 1000000);
```

Comment:
nit: `(e->ts_us - start_ts) / 1000000.0` is equivalent (if not even more precise), but IMO cleaner way to write the same

Reply:
Fixed, thanks!